### ISSUE TITLE
Implement running end-to-end system & polish scaffolding

### ISSUE THREAD

**Issue by pospi** [holo-rea/holo-rea#14]
This issue has been filed to record the outstanding cleanup jobs remaining on https://github.com/holo-rea/holo-rea/pull/13. As you can see there is a lot of tidy-up to do... my goal is to try to keep things moving and avoid a large merge conflict or bottleneck in aligning what we've already done (particularly as I have low bandwidth for development in the coming week).

Though the PR is still rough around the edges the framework, some patterns and at least the locations of files appear to be coalescing and I don't want to waste an opportunity for us to be working within the same collaboration space & polishing out the same details- I think things are stable enough for both of us to be working on top of. For completeness and so you can get a sense of what still needs to be tidied up, here's an exhaustive list of what still needs doing-

code cleanup:

- [ ] move event / commitment zome I/O structs into `vf_core`, abstract out the pattern
    - [ ] use macros to avoid repetition between CREATE, UPDATE, entry (storage) & response (output) structures
- [ ] rename & potentially split "main" zomes, restructure around areas of function
- `vf-graphql` integration:
    - [x] implement build process for reference schema (see https://github.com/valueflows/vf-graphql/issues/32)
        - [x] move schema instantiation logic to `vf-graphql-holochain` package and no longer depend on Webpack for bundling (remove related `raw.macro` & hack in `postinstall.sh` as well)
    - [x] fix scalar types setup in schema resolvers config
    - [ ] remove temporary git submodule from repository, publish module to NPM & reference as separate node package

documentation:

- [ ] document language around "requests", "records", "entries", "base entries" & "links" (perhaps equivalent to docs for `vf_core` record macros)
- [ ] document files, methods & modules to Rust standards
- [x] docs for `example` GraphiQL explorer
- [x] finish scaffolding and documenting `modules/vf-graphql-holochain`

testing:

- [ ] unit tests as appropriate
- integration tests (awaiting bridging support in nodejs conductor)
    - [ ] create event
    - [ ] read event (incl. fulfillment field)
    - [ ] read fulfillments
    - [ ] create event, link fulfillment, modify event, link fulfillment, read fulfillments; validate length = 2

implementation:

- [ ] implement [fixed identity pattern](https://chat.holochain.org/appsup/pl/5qza61nwn3grxgepzb8yhowg3c) for records
- [x] update address type aliases to use newtype pattern to avoid accidental field mismatches
- [x] wire up `Action` fields using static action builtins (note- some `entry_address` wrapping still needed)
- [ ] improve error handling around complex operations
    - [ ] For critical writes (eg. the base entry for an *'entry + links'* record structure), handlers should panic before attempting any other operations.
    - [ ] For parallel writes (for example, linking a set of `EconomicEvent`s as `fulfillment`s of a `Commitment`), handlers should pass back `ZomeApiResult` objects and leave the context handling to the caller. This would usually be done towards the end of the callback handler by partitioning out all the `Err` responses from sub-operations and returning them in the API response object. You could also do things like pass the errors into some bridged "error logging" Holochain DHT.
- [ ] include field name whitelist parameters within every request to avoid reading links where unnecessary
- [ ] expand `fulfillment` test implementation to include the intermediary record struct and abstract out a pattern for 

My goal is to keep working to tick many of these off; things that are to be left for later will be logged as separate issues. @sqykly please let me know if there is anything here you think should be completed before merging the PR- in other words, that you feel blocks you from being able to continue work after merging #13 into what you're doing. Other than that, **I definitely think most of this should be taken care of before proceeding beyond event / fulfillment / commitment**, as I want to ensure we've landed on optimal patterns and laid a good foundation for code best-practises before we continue work elsewhere.

There are also some assumptions in my tasks which might require unpacking, please query anything which doesn't make sense. Perhaps this can become the new "optimal architecture" thread, since it seems pretty clear that `read` & `create` are mostly solved problems and the only thing we may still have to discuss is `update` and how that fits in to the project scaffolding I've already created.

There are a couple of other items that can't be tied off just yet until the Holochain core & HDK catch up (CC @philipbeadle @thedavidmeister @willemolding)-

- Implement bindings into the *capabilities register* and *bridge registration callback* to set up inter-network data operations more flexibly.
    - Allow multiple DNAs to interconnect as the "same" network (2 separate observation DNAs referencing a single planning DNA, for example)
- Integration tests to proof cross-DNA functionality

**Comment by pospi** [comment]
I'm probably going to hit pause on this for now. Have setup a nice streamlined developer experience for the frontend (JavaScript) layers and proven that the project structure, commands & scaffolding all work as intended and can fit whatever patterns end up taking shape within the DNA code.

The rest of the jobs on this list (abstracting away Rust code patterns, documenting the Rust layer, implementing DNA functionality tests & playing with implementation details) don't feel worth pursuing while @sqykly has requested a hold on development and time to explore patterns. I don't want to create friction when it comes time to discuss those patterns and integrate code, which is what it feels like the outcome will be if I go too much further with things.

The two exceptions might be these jobs, which I'll play with if I have time before our next sync call:

- update `address` type aliases to use newtype pattern to avoid accidental field mismatches
- wire up `Action` fields using static action builtins

**Comment by pospi** [comment]
Closing this as the work is pretty much covered by subsequent issues now.
